routing-daemon/lib/openshift/routing/daemon.rb

```diff
@@ -333,65 +349,75 @@ def delete_application app_name, namespace
-      create_application app_name, namespace if @lb_controller.pools[pool_name].nil?
-      unless (types & @endpoint_types).empty?
-        @logger.info "Adding new member #{gear_host}:#{gear_port} to pool #{pool_name}"
-        @lb_controller.pools[pool_name].add_member gear_host, gear_port.to_i
-      else
-        @logger.info "Ignoring endpoint with types #{types.join(',')}"
+      with_pool app_name, namespace do |pool|
+        unless (types & @endpoint_types).empty?
+          @logger.info "Adding new member #{gear_host}:#{gear_port} to pool #{pool.name}"
+          pool.add_member gear_host, gear_port.to_i
+        else
+          @logger.info "Ignoring endpoint with types #{types.join(',')}"
+        end
       end
     end
 
     def remove_endpoint app_name, namespace, gear_host, gear_port
-      pool_name = generate_pool_name app_name, namespace
-      member = gear_host + ':' + gear_port.to_s
-      # The remove_endpoint notification does not include a "types" field, so we
-      # cannot simply filter out endpoint types that we don't care about.
-      # Furthermore, the daemon by design does not store state itself, so we
-      # cannot check keep track of endpoints that we did not add to the load
-      # balancer.  All we can do is check whether the endpoint exists and delete
-      # it if it does.
-      if @lb_controller.pools[pool_name].members.include?(member)
-        @logger.info "Deleting member #{member} from pool #{pool_name}"
-        @lb_controller.pools[pool_name].delete_member gear_host, gear_port.to_i
-      else
-        @logger.info "No member #{member} exists in pool #{pool_name}; ignoring"
+      with_pool app_name, namespace, false do |pool|
+        member = gear_host + ':' + gear_port.to_s
+        if not pool.members.include?(member)
+          @logger.info "Pool member #{member} not found in pool #{pool.name}; resynching pool..."
+          pool.synch
+        end
+
+        # The remove_endpoint notification does not include a "types" field, so we
+        # cannot simply filter out endpoint types that we don't care about.
+        # Furthermore, the daemon by design does not store state itself, so we
+        # cannot check keep track of endpoints that we did not add to the load
+        # balancer.  All we can do is check whether the endpoint exists and delete
+        # it if it does.
+        if pool.members.include?(member)
+          @logger.info "Deleting member #{member} from pool #{pool.name}"
+          pool.delete_member gear_host, gear_port.to_i
+        else
+          @logger.info "No member #{member} exists in pool #{pool.name}; ignoring"
+        end
       end
     end
 
     def add_alias app_name, namespace, alias_str
-      pool_name = generate_pool_name app_name, namespace
-      @logger.info "Adding new alias #{alias_str} to pool #{pool_name}"
-      @lb_controller.pools[pool_name].add_alias alias_str
+      with_pool app_name, namespace do |pool|
```

Comment:
Should this be passing `create_if_nil=false`, since it was not creating the application previously?